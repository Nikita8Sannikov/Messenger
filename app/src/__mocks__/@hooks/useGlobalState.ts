export const mockUseAppSelector = jest.fn();
export const mockUseAppDispatch = jest.fn();