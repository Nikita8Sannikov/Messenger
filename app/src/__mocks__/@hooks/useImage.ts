const useImage = (src: string | null): string => {
	return src || "/assets/images/noPhoto.jpg";
};

export default useImage; 