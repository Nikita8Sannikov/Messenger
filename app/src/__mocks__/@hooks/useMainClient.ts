const mockMainApi = {
	getAnotherUser: jest.fn(),
	signIn: jest.fn(),
	signUp: jest.fn(),
	logout: jest.fn(),
	uploadAvatarAuth: jest.fn(),
	getFriendsNotification: jest.fn(),
	getMessageNotification: jest.fn(),
	openFile: jest.fn(),
};

const useMainClient = jest.fn().mockReturnValue({ 
	mainApi: mockMainApi,
	getProfile: jest.fn(),
	existProfile: jest.fn(),
	lifeTimeExpire: jest.fn(),
	removeProfile: jest.fn(),
	downloadLogFile: jest.fn(),
});

export default useMainClient;
export { mockMainApi }; 