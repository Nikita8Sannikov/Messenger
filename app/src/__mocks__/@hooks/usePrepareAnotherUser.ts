export interface IUsePrepareAnotherUser {
  isLoading: boolean;
  isExistProfile: boolean;
}

const mockUsePrepareAnotherUser = (): IUsePrepareAnotherUser => ({
	isLoading: false,
	isExistProfile: true,
});

export default mockUsePrepareAnotherUser; 