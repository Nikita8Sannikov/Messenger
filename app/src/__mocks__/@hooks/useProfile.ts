import { type IPhoto } from "@custom-types/models.types";

const mockNotificationSettingsService = {
	soundEnabled: false,
	messageSound: false,
	friendRequestSound: false,
	changeSoundNotifications: jest.fn(),
};

const mockPhotosService = {
	photos: [] as IPhoto[],
	count: 0,
	getAllPhotos: jest.fn(),
	onClickPhoto: jest.fn(),
	addPhotoFromAvatar: jest.fn(),
	addPhotos: jest.fn(),
	deletePhoto: jest.fn(),
};

const mockUserDetailsService = {
	updateDetails: jest.fn(),
	syncUserDetails: jest.fn(),
	getFriendsText: jest.fn(),
	getSubscribersText: jest.fn(),
	getPhotosText: jest.fn(),
	getAudiosText: jest.fn(),
	getVideosText: jest.fn(),
};

const mockUserService = {
	photosService: mockPhotosService,
	detailsService: mockUserDetailsService,
	settingsService: mockNotificationSettingsService,
	id: "",
	firstName: "",
	thirdName: "",
	phone: "",
	email: "",
	fullName: "",
	avatarUrl: "",
	avatarCreateDate: "",
	updateUser: jest.fn(),
	syncInfo: jest.fn(),
	changeAvatar: jest.fn(),
};

const mockProfileService = {
	isMe: false,
	userService: mockUserService,
	photosService: mockPhotosService,
	deleteAccount: jest.fn(),
	onSetAvatar: jest.fn(),
	onDeleteAvatar: jest.fn(),
	onClickAvatar: jest.fn(),
	getFriends: jest.fn(),
	editInfo: jest.fn(),
};

const useProfile = jest.fn().mockReturnValue(mockProfileService);

export default useProfile;
export { 
	mockProfileService, 
	mockUserService, 
	mockPhotosService, 
	mockUserDetailsService, 
	mockNotificationSettingsService, 
};