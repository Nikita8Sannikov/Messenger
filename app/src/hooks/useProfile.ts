import useMainClient from "@hooks/useMainClient";

// Возврат сущности "Профиль" конкретного пользователя по переданному идентификатору
export default function useProfile(userId?: string) {
	const mainClient = useMainClient();
	return mainClient.getProfile(userId);
}
