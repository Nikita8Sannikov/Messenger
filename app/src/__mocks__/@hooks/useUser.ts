import { mockUserService } from "./useProfile";

const useUser = jest.fn().mockReturnValue(mockUserService);

export default useUser; 