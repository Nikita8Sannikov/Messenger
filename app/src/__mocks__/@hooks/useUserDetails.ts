import { mockUserDetailsService } from "./useProfile";

const useUserDetails = jest.fn().mockReturnValue(mockUserDetailsService);

export default useUserDetails; 