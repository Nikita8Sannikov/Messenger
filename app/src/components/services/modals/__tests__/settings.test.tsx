import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import ModalWithSettings from "@components/services/modals/settings";
import { mockNotificationSettingsService } from "../../../../__mocks__/@hooks/useProfile";
import mockProfileStore from "../../../../__mocks__/@store/profile";
import mockUIStore from "../../../../__mocks__/@store/ui";

describe("ModalWithSettings", () => {
	beforeEach(() => {
		jest.clearAllMocks();

		const store = mockUIStore.getState();
		store.settingsModal = true;
		store.saveSettingsModal = false;
		store.setSettingsModal = jest.fn();
		store.setConfirmModal = jest.fn();

		const profileStore = mockProfileStore.getState();
		profileStore.isDeleteAccountLoading = false;
	});

	describe("Form initialization", () => {
		it("should initialize form with settings service values when modal is open", () => {
			render(<ModalWithSettings />);

			const soundSwitch = screen.getByLabelText("All sound notifications");
			const messageSwitch = screen.getByLabelText("Sound notifications when receiving a message");
			const friendRequestSwitch = screen.getByLabelText("Sound notifications when receiving a friend's request");

			expect(soundSwitch).not.toBeChecked();
			expect(messageSwitch).not.toBeChecked();
			expect(friendRequestSwitch).not.toBeChecked();
		});

		it("should not initialize form when modal is closed", () => {
			const store = mockUIStore.getState();
			store.settingsModal = false;

			const { container } = render(<ModalWithSettings />);
			expect(container).toBeEmptyDOMElement();
		});

		it("should not initialize form when settings service is not available", () => {
			const { container } = render(<ModalWithSettings />);
			expect(container).toBeEmptyDOMElement();
		});

		it("should update form when settings service values change", async () => {
			const { rerender } = render(<ModalWithSettings />);

			mockNotificationSettingsService.soundEnabled = false;
			mockNotificationSettingsService.messageSound = false;
			mockNotificationSettingsService.friendRequestSound = false;

			rerender(<ModalWithSettings />);

			const soundSwitch = screen.getByLabelText("All sound notifications");
			const messageSwitch = screen.getByLabelText("Sound notifications when receiving a message");
			const friendRequestSwitch = screen.getByLabelText("Sound notifications when receiving a friend's request");

			expect(soundSwitch).not.toBeChecked();
			expect(messageSwitch).not.toBeChecked();
			expect(friendRequestSwitch).not.toBeChecked();
		});
	});

	it("should match snapshot when modal is open", () => {
		const { container } = render(<ModalWithSettings />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot when modal is closed", () => {
		const store = mockUIStore.getState();
		store.settingsModal = false;
		store.saveSettingsModal = false;
		store.setSettingsModal = jest.fn();
		store.setConfirmModal = jest.fn();

		const { container } = render(<ModalWithSettings />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot when saving", () => {
		const store = mockUIStore.getState();
		store.settingsModal = true;
		store.saveSettingsModal = true;
		store.setSettingsModal = jest.fn();
		store.setConfirmModal = jest.fn();

		const { container } = render(<ModalWithSettings />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot when deleting account", () => {
		const profileStore = mockProfileStore.getState();
		profileStore.isDeleteAccountLoading = true;

		const { container } = render(<ModalWithSettings />);
		expect(container).toMatchSnapshot();
	});

	it("should not render when modal is closed", () => {
		const store = mockUIStore.getState();
		store.settingsModal = false;
		store.saveSettingsModal = false;
		store.setSettingsModal = jest.fn();
		store.setConfirmModal = jest.fn();

		const { container } = render(<ModalWithSettings />);
		expect(container).toBeEmptyDOMElement();
	});

	it("should render with initial settings", () => {
		render(<ModalWithSettings />);

		expect(screen.getByText("App settings")).toBeInTheDocument();
		expect(screen.getByText("Deleting your profile.")).toBeInTheDocument();
		expect(screen.getByText("Save")).toBeInTheDocument();
		expect(screen.getByText("Cancel")).toBeInTheDocument();
	});

	it("should handle form changes", async () => {
		render(<ModalWithSettings />);

		const soundSwitch = screen.getByLabelText("All sound notifications");
		fireEvent.click(soundSwitch);

		const saveButton = screen.getByText("Save");
		fireEvent.click(saveButton);

		await waitFor(() => {
			expect(mockNotificationSettingsService.changeSoundNotifications).toHaveBeenCalledWith({
				soundEnabled: true,
				messageSound: false,
				friendRequestSound: false,
			});
		});
	});

	it("should handle save button click", async () => {
		render(<ModalWithSettings />);

		const saveButton = screen.getByText("Save");
		fireEvent.click(saveButton);

		await waitFor(() => {
			expect(mockNotificationSettingsService.changeSoundNotifications).toHaveBeenCalledWith({
				soundEnabled: false,
				messageSound: false,
				friendRequestSound: false,
			});
		});
	});

	it("should handle delete account button click", () => {
		render(<ModalWithSettings />);

		const deleteButton = screen.getByText("Delete profile");
		fireEvent.click(deleteButton);

		expect(mockUIStore.getState().setConfirmModal).toHaveBeenCalledWith({
			text: "Do you really want to delete your account?",
			btnActionTitle: "Delete",
			cb: expect.any(Function),
		});
	});

	it("should handle cancel button click", () => {
		render(<ModalWithSettings />);

		const cancelButton = screen.getByText("Cancel");
		fireEvent.click(cancelButton);

		expect(mockUIStore.getState().setSettingsModal).toHaveBeenCalledWith(false);
	});

	it("should show loading state when saving", () => {
		const store = mockUIStore.getState();
		store.settingsModal = true;
		store.saveSettingsModal = true;
		store.setSettingsModal = jest.fn();
		store.setConfirmModal = jest.fn();

		render(<ModalWithSettings />);

		const saveButton = screen.getByText("Save");
		expect(saveButton).toHaveClass("Mui-disabled");
	});

	it("should show loading state when deleting account", () => {
		const profileStore = mockProfileStore.getState();
		profileStore.isDeleteAccountLoading = true;

		render(<ModalWithSettings />);

		const deleteButton = screen.getByText("Delete profile");
		expect(deleteButton).toHaveClass("Mui-disabled");
	});
});
