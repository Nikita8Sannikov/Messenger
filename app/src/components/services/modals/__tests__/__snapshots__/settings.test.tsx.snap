// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWithSettings should match snapshot when deleting account 1`] = `<div />`;

exports[`ModalWithSettings should match snapshot when modal is closed 1`] = `<div />`;

exports[`ModalWithSettings should match snapshot when modal is open 1`] = `<div />`;

exports[`ModalWithSettings should match snapshot when saving 1`] = `<div />`;
